ttnn/cpp/ttnn/operations/ccl/kernel_common/worker_sync_utils.hpp

```diff
@@ -151,3 +196,157 @@ FORCE_INLINE void advance_start_page_idx(
     }
 
 }
+
+
+struct MatmulOpReceiver {
+    uint32_t num_tensor_slices;
+    uint32_t num_directions = 2; // ASSUMPTION: Always 2 directions
+
+    bool wait_for_op_signal;
+    uint32_t num_transfers;
+    uint32_t ring_size;
+    uint32_t tensor_slice_shape_width; // In tiles
+    uint32_t output_page_offset;
+    uint32_t last_output_page_offset;
+
+    uint32_t num_blocks;
+    uint32_t num_blocks_per_slice;
+
+    // Used to track internal state
+    uint32_t ring_idxs[2];
+    uint32_t start_page_idxs[2];
+    bool is_clockwise_dirs[2];
+    volatile tt_l1_ptr uint32_t* signal_op_semaphore_addr_ptrs[2];
+    uint32_t curr_dir;
+    uint32_t curr_transfer_idx;
+
+
+    bool initialized = false;
+
+    MatmulOpReceiver() {}
+
+    MatmulOpReceiver(
+        bool wait_for_op_signal,
+        uint32_t& rt_args_idx,
+        uint32_t num_blocks,
+        uint32_t tiles_per_block // Across the same dimension as tensor_slice_shape_width
+    ) : wait_for_op_signal(wait_for_op_signal),
+        num_blocks(num_blocks)
+    {
+
+        // Runtime args
+        this->num_transfers = get_arg_val<uint32_t>(rt_args_idx++);
+        this->ring_size = get_arg_val<uint32_t>(rt_args_idx++);
+        uint32_t start_ring_index = get_arg_val<uint32_t>(rt_args_idx++);
+        this->tensor_slice_shape_width = get_arg_val<uint32_t>(rt_args_idx++);
+        this->output_page_offset = get_arg_val<uint32_t>(rt_args_idx++);
+        this->last_output_page_offset = get_arg_val<uint32_t>(rt_args_idx++);
+        uint32_t is_clockwise_direction = get_arg_val<uint32_t>(rt_args_idx++);
+
+        if (this->wait_for_op_signal) {
+            this->signal_op_semaphore_addr_ptrs[0] =
+                reinterpret_cast<volatile tt_l1_ptr uint32_t*>(get_semaphore(get_arg_val<uint32_t>(rt_args_idx++)));
+            this->signal_op_semaphore_addr_ptrs[1] =
+                reinterpret_cast<volatile tt_l1_ptr uint32_t*>(get_semaphore(get_arg_val<uint32_t>(rt_args_idx++)));
+        }
+
+        this->num_tensor_slices = this->num_transfers * this->num_directions;
+
+        // Setup internal states for bi-direction
+        this->ring_idxs[0] = start_ring_index;
+        this->ring_idxs[1] = start_ring_index;
+
+        this->start_page_idxs[0] = this->ring_idxs[0] * this->output_page_offset;
+        this->start_page_idxs[1] = this->ring_idxs[1] * this->output_page_offset;
+
+        this->is_clockwise_dirs[0] = is_clockwise_direction;
+        this->is_clockwise_dirs[1] = !is_clockwise_direction;
+
+        this->num_blocks_per_slice = this->tensor_slice_shape_width / tiles_per_block;
+        ASSERT(this->num_tensor_slices * this->num_blocks_per_slice == this->num_blocks);
+
+        this->curr_dir = is_clockwise_direction ? 1 : 0; // Anti-clockwise direction is the first since it has local slice
+        this->curr_transfer_idx = 0;
+
+        this->initialized = true;
```

Comment:
I don't this we need this is initialized bool anymore now that you've consolidated all the initialization into the constructor. Is that correct?

Reply:
there's still an empty default constructor though, which means that it's possible to create an object that's not initialized